### ISSUE TITLE
Stack blown with circular reference

### ISSUE THREAD

**Issue by nonara** [TehShrike/deepmerge#207]
In a circular reference scenario, it gets into an infinite recursion scenario, which blows the stack.

In these scenarios, using a map to track certain objects (ie. key: previous object, value: merged object) can mitigate the issue.

Reproduction:

```js
a = { self: null };
a.self = a;
common_1.deepMerge.all([ a, { b: 2 } ])
```

**Comment by TehShrike** [comment]
This would probably best be solved by a custom `isMergeableObject` that keeps track of what it's compared to, and returns `false` if it's an object that has been seen before.

I would be open to linking to such an implementation, or perhaps even putting it into the readme for people to copy/paste.

**Comment by nonara** [comment]
Updated my original comment with a strategy.

Because of the other issue I posted, this isn't possible for me at the moment. However, I would suggest in this case mitigating. Circular reference issues are bound to happen in many scenarios. This is an issue that we have to deal with in compilers & parsers very often.

In this case, anything with a reference to a node higher up the chain is going to blow the stack, which in general should be avoided. The proposed strategy above is the general approach which mitigates the issue or throws an appropriate error message if it can't.

Are you open to a PR?

**Comment by nonara** [comment]
> This would probably best be solved by a custom `isMergeableObject` that keeps track of what it's compared to, and returns `false` if it's an object that has been seen before.

Using seen would cause false positives in some cases.

Example:

- Root
   - Child1
       - Grandchild 1
   - Child2
       - Grandchild 2

If we're walking Child2, Child1 and Grandchild1 would have been 'seen', but because they're not in the ancestral lineage, they can be merged.

That's why we need a direct lineage map to be passed to the recursive function. Without a third parameter, an `isMergeableObject` implementation couldn't do this.

But again, this isn't too difficult to solve and could be taken care of with a few extra lines. Let me know if you'd consider a PR.

**Comment by TehShrike** [comment]
actually, on reflection, this is a much harder problem that I don't really want to try to solve here – merging two trees makes sense, but merging two cyclical graphs doesn't make sense – what would you even expect the output to be if you tried to merge

```js
const a = { child: { grandchild: { tiny_baby: 'hi } } }

const b = { child: { } }
b.child.grandchild = b

merge(a, b)
```

That seems nondeterministic, right?

I'm inclined to say "the solution is to eliminate cycles in your objects before trying to merge them"

**Comment by nonara** [comment]
Effectively, that becomes:

```js
const a = { child: { grandChild: b } };
const b = { child: { } }
```

In this case, we understand that `b` is a reference to a parent node that has been merged (`a` and `b` are now a new merged node). Because this would result in a circular reference and we cannot merge, the expected result would be to output a reference to the corresponding merged node. In this case, the `b` would be a reference to the root.

```js
const result = { child: { grandChild: result }
```

**Comment by nonara** [comment]
Effectively, that becomes:

```js
const a = { child: { grandchild: { tiny_baby: 'hi' } };
const b = { child: { grandchild: b } };
```

You're correct that in this case this cannot be merged. Effectively here, if we were to assign a pseudo-AST to it, we have a `Node` and a `ParentReferenceNode`. This would be a non-mergeable scenario. In my proposed method above, I propose allowing a custom handler function for dealing with circular merge situations, and if one doesn't exist, we use default behaviour.

> 3. Handle circular reference
>    - If there is nothing to merge, set output value to the value from our map
>    - If there is something to merge and no custom logic exists, throw a circular reference error

In your example, it would either pass off the conflict to a handler function, or if none is given, it would throw a circular reference error.

**Comment by nonara** [comment]
As far as whether it should be solved I have two thoughts:

#### 1. It's generally a good idea to ensure that recursive functions cannot ever hit an infinite recursion scenario.

That in mind, putting in some logic to to track this is good practice. And if you're going to do that, the extra bits to allow handling those scenarios isn't too much extra and would help in making it so only those scenarios that should throw an exception do. (ie, there's a reference to a parent but nothing to merge with it, we can just keep the reference as a reference to its merged counterpart)

#### 2. How often will two graphs be merged?

It depends on the context. This is very common in some areas of CS, and not as much in others. Ultimately, it's your call. Not supporting it will limit the scopes in which the library can be used, where adding the ability to handle circular references will certainly broaden that and help a lot of people.

At the very least, it will reduce bug reports when people find themselves hitting circular ref issues.

Again, the rules are pretty simple (pseudo-AST again)

1. If one property is a `ParentReferenceNode`, and there is no node to merge 
   - Call a handler function if set to determine output
   - (Default if no handler) Output a reference to the corresponding merged parent node

2. If one property is a `ParentReferenceNode` and there is another node to merge 
   - Call a handler function if set to determine output
   - (Default if no handler) Throw a circular reference exception

The default behaviour can be changed, but this prevents stack being blown and allows users to have control over circular issues, which will allow the library to be used in more contexts.

**Comment by TehShrike** [comment]
> It's generally a good idea to ensure that recursive functions cannot ever hit an infinite recursion scenario

I'm amenable to this.  If we added any code to address this, my preferred solution would be to throw an error with a clear message as soon as a cycle was detected.

Merging data structures in JS is complicated enough already, I don't want to make my life the hell that it would become by trying to solve general CS "merging cyclical graphs" problems.

**Comment by nonara** [comment]
Makes sense. How about this:

1. Track circular refs
2. If `customMerge` is set > Pass a third parameter to the `customMerge` function called `circularReferences`. 
3. If `customMerge` is not set or it returns undefined > The default behaviour throws a circular ref exception, per your preference

`circularReferences` would be either by `undefined` if neither a or b are circular, or it will be a simple map if either is.
The map would be `<sourceReference, targetReference>` (see below for example)

Simple, non-breaking, and still allows us hell-dwellers to do our thing, as needed. 😄 

#### Example

The following is how I could implement the behaviour I proposed earlier using this extra param:

```js
let a, b;
a = { child: { parent: a.child } }
b = { hi: 1 }

merge(a, b, { 
  customMerge: () => (a, b, circularReferences) => {
    if (circularReferences && (a === undefined || b === undefined)) 
      return [ ...circularReferences.values() ][0]; // The value will be a reference to mergedObj.child
  }
});
```

If you're down with that, let me know and I'll do the PR.

**Comment by martin-braun** [comment]
deepmerge should not blow up when merging with circular references. I'd love to see this getting implemented.

**Comment by TehShrike** [comment]
> deepmerge should not blow up when merging with circular references. I'd love to see this getting implemented.

Please read the thread above – I'm open to throwing a clear error when a cycle is detected, but I do not see an obvious way to return sensical data when merging a cycle.

**Comment by RebeccaStevens** [comment]
I think the default behavior should be throwing an error. If custom merge function is provided then that function can determine how to handle cycles.

**Comment by nonara** [comment]
> I think the default behavior should be throwing an error. If custom merge function is provided then that function can determine how to handle cycles.

Agreed. My proposal above should meet those specifications. I'd still willing to do the PR.

**Comment by martin-braun** [comment]
@TehShrike I agree that handling circular references is not an easy task, but I think a good starting point is to inspect [solutions on stringifying circular references](https://stackoverflow.com/questions/11616630/how-can-i-print-a-circular-structure-in-a-json-like-format). I think a similar approach could handle circular references here, so it would allow to deep merge objects with circular references without throwing an error.

This implementation would give this package a superior status when it comes to deepmerge-solutions. 

> I think the default behavior should be throwing an error. If custom merge function is provided then that function can determine how to handle cycles.

I think this is a first good logical step, though.

**Comment by nonara** [comment]
Unless I'm misunderstanding you, stringifying would not work as it would still have a circular issue.

This sort of thing is not as complicated as it may seem. It's quite common in AST-like scenarios. I apologize if I've not explained the proposed solution clearly.

To clarify, you simply maintain a linear set of the parent references during the walk, then pass that set to the custom merge function, if present. The custom merge function can determine what to do if `a` or `b` are in the passed set. If no custom merge function is present, or the function returns a value that is still a circular reference, it will throw a circular error.

This allows for fast-exit throw behaviour as the default while also providing for custom handling if the user desires to.

**Comment by martin-braun** [comment]
> Unless I'm misunderstanding you, stringifying would not work as it would still have a circular issue.

@nonara Unfortunately, you misunderstood me. I know that stringify is not a solution to this problem, I was just saying that it *has* the same problem and it can be worked around. I was just making a point that it's not impossible to fix this, properly.

> This sort of thing is not as complicated as it may seem. It's quite common in AST-like scenarios. I apologize if I've not explained the proposed solution clearly.
> 
> To clarify, you simply maintain a linear set of parent references during the walk, then pass that set to the custom merge function, if present. The custom merge function can determine what to do if `a` or `b` are in the passed set. If no custom merge function is present, or the function returns a value that is still a circular reference, it will throw a circular error.
> 
> This allows for fast-exit throw behaviour as the default while also providing for custom handling if the user desires to.

I get your point, but why not implementing that _custom function_ that can handle circular references into the lib in the first place? The lib already has a fixed behavior when same keys have different values, so it simply needs to maintain references and set them before causing any circular references, as you described very well.

So, your reproduction sample

```js
a = { self: null };
a.self = a;
common_1.deepMerge.all([ a, { b: 2 } ])
```

should result in an object that looks like this: 

```json
{
  "self": [object Object],
  "b": 2
}
```

while the root object `a` equals `a.self`. This could work without any errors.
